Data/Primitive/Array/Class.hs

```diff
@@ -0,0 +1,485 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+
+{-|
+Module      : Data.Array
+Description : Unified boxed and unboxed arrays
+Copyright   : (c) Winter, 2017
+License     : BSD3
+Maintainer  : libraries@haskell.org, drkoster@qq.com
+Stability   : experimental
+Portability : non-portable
+
+Unified unboxed and boxed array operations using functional dependencies.
+
+All operations are NOT bound checked, if you need checked operations please use "Data.Array.Checked".
+It exports exactly same APIs so that you can switch between without pain.
+
+-}
+
+module Data.Primitive.Array.Class (
+  -- * Arr typeclass
+    Arr(..)
+  , RealWorld
+  -- * Boxed array type
+  , Array(..)
+  , MutableArray(..)
+  , SmallArray(..)
+  , SmallMutableArray(..)
+  , uninitialized
+  -- * Primitive array type
+  , PrimArray(..)
+  , MutablePrimArray(..)
+  , Prim(..)
+  , newPinnedPrimArray, newAlignedPinnedPrimArray
+  , copyPrimArrayToPtr, copyMutablePrimArrayToPtr, copyMutablePrimArrayFromPtr
+  -- * Unlifted array type
+  , UnliftedArray(..)
+  , MutableUnliftedArray(..)
+  , PrimUnlifted(..)
+  -- * The 'ArrayException' type
+  , ArrayException(..)
+  ) where
+
+import Data.Primitive.Types
+import Control.Monad.Primitive
+import Control.Exception (ArrayException(..), throw)
+import Data.Primitive.PrimArray
+import Data.Primitive.Array
+import Data.Primitive.SmallArray
+import Data.Primitive.UnliftedArray
+import GHC.ST
+import GHC.Prim
+import Data.Primitive.Internal.Compat ( isTrue# )
+
+#if (__GLASGOW_HASKELL__ >= 710)
+#define HAVE_SMALL_ARRAY 1
+#endif
+
+-- | Bottom value (@throw ('UndefinedElement' "Data.Array.uninitialized")@)
+-- for initialize new boxed array('Array', 'SmallArray'..).
+--
+uninitialized :: a
+uninitialized = throw (UndefinedElement "Data.Primitive.Array.Class.uninitialized")
+
+-- | A typeclass to unify box & unboxed, mutable & immutable array operations.
+--
+-- Most of these functions simply wrap their primitive counterpart. If there are no primitive ones,
+-- the method is polyfilled using other operations to get the same semantics.
+--
+-- One exception is that 'shrinkMutableArr' only performs resizing on 'PrimArray' because
+-- the RTS only supports shrinking byte arrays. On other array types, 'shrinkMutableArr' will do nothing.
+--
+class Arr (marr :: * -> * -> *) (arr :: * -> * ) a | arr -> marr, marr -> arr where
```

Comment:
My immediate concern when seeing this interface is that this doesn't allow the same interface to work For ArrayArray too! Then we can't have a nice interface for arrays of unlifted values!